browser/ui/views/brave_actions/brave_rewards_action_view.cc

```diff
@@ -0,0 +1,417 @@
+  if (selection.active_tab_changed()) {
+    UpdateTabHelper(selection.new_contents);
+  }
+}
+
+void BraveRewardsActionView::OnPublisherForTabUpdated(
+    const std::string& publisher_id) {
+  publisher_registered_ = {publisher_id, false};
+  bool status_updating = UpdatePublisherStatus();
+  if (!status_updating) {
+    Update();
+  }
+}
+
+bool BraveRewardsActionView::OpenRewardsPanel(
+    const brave_rewards::mojom::RewardsPanelArgs& args) {
+  if (!IsPanelOpen()) {
+    ToggleRewardsPanel();
+  }
+  return true;
+}
+
+void BraveRewardsActionView::OnPublisherRegistryUpdated() {
+  UpdatePublisherStatus();
+}
+
+void BraveRewardsActionView::OnPublisherUpdated(
+    const std::string& publisher_id) {
+  if (publisher_id == std::get<std::string>(publisher_registered_)) {
+    UpdatePublisherStatus();
+  }
+}
+
+void BraveRewardsActionView::OnNotificationAdded(
+    RewardsNotificationService* service,
+    const RewardsNotificationService::RewardsNotification& notification) {
+  Update();
+}
+
+void BraveRewardsActionView::OnNotificationDeleted(
+    RewardsNotificationService* service,
+    const RewardsNotificationService::RewardsNotification& notification) {
+  Update();
+}
+
+void BraveRewardsActionView::OnButtonPressed() {
+  // If we are opening the Rewards panel, use `RewardsPanelCoordinator` to open
+  // it so that the panel arguments will be correctly set.
+  if (!IsPanelOpen() && panel_coordinator_) {
+    panel_coordinator_->OpenRewardsPanel();
+    return;
+  }
+
+  ToggleRewardsPanel();
+}
+
+void BraveRewardsActionView::OnPreferencesChanged(const std::string& key) {
+  Update();
+}
+
+content::WebContents* BraveRewardsActionView::GetActiveWebContents() {
+  return browser_->tab_strip_model()->GetActiveWebContents();
+}
+
+brave_rewards::RewardsService* BraveRewardsActionView::GetRewardsService() {
+  return RewardsServiceFactory::GetForProfile(browser_->profile());
+}
+
+brave_rewards::RewardsNotificationService*
+BraveRewardsActionView::GetNotificationService() {
+  if (auto* rewards_service = GetRewardsService()) {
+    return rewards_service->GetNotificationService();
+  }
+  return nullptr;
+}
+
+bool BraveRewardsActionView::IsPanelOpen() {
+  return bubble_observation_.IsObserving();
+}
+
+void BraveRewardsActionView::ToggleRewardsPanel() {
+  if (IsPanelOpen()) {
+    DCHECK(bubble_manager_);
+    bubble_manager_->CloseBubble();
+    return;
+  }
+
+  // Clear the default-on-start badge text when the user opens the panel.
+  auto* prefs = browser_->profile()->GetPrefs();
+  prefs->SetString(brave_rewards::prefs::kBadgeText, "");
+
+  // Create a new instance of the bubble manager. Note that this disables the
+  // `WebContents` caching mechanism of `WebUIBubbleManager`. To support
+  // caching, the front end will need to be able to receive a notification when
+  // the panel is requested and obtain the panel arguments as approprite. The
+  // front-end should not rely on document visibility changes for this purpose.
```

Comment:
This is a shame because:
1. The panel does not seem very quick to open, and I observe a couple of flashes between black and white (depending on active OS theme)
2. The WebContents _is_ being cached according to chrome://inspect and just not being used. So we're using resources even though no benefit will be realized. At least it's only kept in memory for a minute or so.

I wonder if we can at least create a follow-up issue so we know what the plan is to get this working, if there's a scope reason not to include it in this PR?

Seems like `RewardsPanelHandler` could observe when the coordinator wants to open / change arguments and can dispatch a callback to the frontend with the new arguments?

Reply:
>  we're using resources even though no benefit will be realized

This is a good argument for moving back to an Observer-style API and supporting the caching. Would it be OK to implement that in a follow-up?

Reply:
sure!